### ISSUE TITLE
Speed issues

### ISSUE THREAD

**Issue by olmobrutall** [s-panferov/awesome-typescript-loader#118]
Hi @s-panferov and company, 

I've been struggling recently improving the build time of a mid-big application using typescript + react + webpack. 

This are my main dependencies: 

``` JS
module.exports = {
    entry: {
        //main: [ "./Scripts/Main.tsx" ],
        "react": ["react", "react-bootstrap", "react-router", "react-widgets", "react-router-bootstrap", "moment", "numbro"],
        "codemirror": ["codemirror"],
        "d3": ["d3"],
    },
    ....
}
```
1. With `main` commented-out webpack takes only **3.9s** to compile the dependencies. 
2. Running `tsc  --outFile` takes **6s**  to compile and bundle my code. 
3. So I was expecting `ts-loader` and `awesome-typescript-loader` to take about **10s**.

Unfortunately It takes between **30s** and **40s**, 3 or 4 times slower than expected. 

I've tried different possibilities, 
- like using `externals` and `noParse`-> no important changes
- stop using `CommonsChunkPlugin`-> no important changes. 
- using `new webpack.optimize.LimitChunkCountPlugin({maxChunks: 15})`  -> about 8% slower.
- removing all the async requires and make one big bundle, without code splitting -> about 10% faster. 
- using `awesome-typescript-loader` instead of `ts-loader` -> about 10% faster. 
- finally using 'ts-loader` with transpileOnly -> _about 80% faster_ now it takes some reasonable \* 12s\* 

But of course this is cheating since `ts-loader` is not doing the same work than tsc was doing.  

For now I will settle on this solution, since I'm compiling TS with noEmit on VisualStudio anyway (I prefer to get the errors there, so I can double-click on them), but I'm curious why webpack is fast, typescript is fast, but webpack + typescript is four times slower :S 

I repeated all the measures twice 

![image](https://cloud.githubusercontent.com/assets/720953/14405549/c8d179f6-fe91-11e5-8f2b-372827ce9347.png)

**Comment by s-panferov** [comment]
@olmobrutall could you please show your tsconfig? 

Another questions:

1) Do you have `exclude` directive in you `tsconfig.json`?
2) Do you have sourceMap enabled? What kind of [`devtool`](https://webpack.github.io/docs/configuration.html#devtool) is enabled in your webpack?
3) What version of TypeScript are you using?
4) What version of NodeJS are you using?

**Comment by olmobrutall** [comment]
Hi @s-panferov,

Thanks for your answer:

```JS
/// <binding />
"use strict";
var path = require('path');
var webpack = require('webpack');
var CleanWebpackPlugin = require('clean-webpack-plugin');
var AssetsPlugin = require('assets-webpack-plugin');
var BeepPlugin = require('webpack-beep-plugin');
var node_modules = path.join(__dirname, "node_modules");


module.exports = {
    entry: {
        main: [ "./Scripts/Main.tsx" ],
        "react": ["react", "react-bootstrap", "react-router", "react-widgets", "react-router-bootstrap", "moment", "numbro"],
        "codemirror": ["codemirror"],
        "d3": ["d3"],
    },
    output: {
        path: path.join(__dirname, "dist"),
        filename: "bundle.[name].[chunkhash].js",
        chunkFilename: "bundle.[name].[chunkhash].js"
    },
    resolve: {
        root: node_modules,
        extensions: ['', '.Webpack.js', '.web.js', '.ts', '.js', '.tsx']     
    },
    resolveLoader: { root: path.join(__dirname, "node_modules") },
    module: {
        //noParse : [
        //    //node_modules + "/react/dist/react.min.js",
        //    node_modules + "/react-dom/dist/react-dom.min.js",
        //    node_modules + "/react-bootstrap/dist/react-bootstrap.min.js",
        //    node_modules + "/react-router/umd/ReactRouter.min.js",
        //    node_modules + "/react-widgets/dist/react-widgets.js"
        //],
        loaders: [
            { test: /\.tsx?$/, loader: 'awesome-typescript-loader' },
            //{ test: /\.tsx?$/, loader: 'ts-loader' },
            //{ test: /\.tsx?$/, loader: 'happypack/loader' },
            //{ test: /\.jsx?$/, loader: "babel-loader" }
            { test: /\.css$/, loader: "style-loader!css-loader" },
            { test: /\.less$/, loader: "style-loader!css-loader!less-loader" },
            { test: /\.gif$/, loader: "url-loader?mimetype=image/png" },
            { test: /\.woff(2)?(\?v=[0-9].[0-9].[0-9])?$/, loader: "url-loader?mimetype=application/font-woff" },
            { test: /\.(ttf|eot|svg)(\?v=[0-9].[0-9].[0-9])?$/, loader: "file-loader?name=[name].[ext]" },
        ]
    },
    plugins: [
        new webpack.optimize.LimitChunkCountPlugin({maxChunks: 15}),
        new webpack.optimize.CommonsChunkPlugin({
            names: ['react', 'codemirror', 'd3'],
            minChunks : Infinity
        }),
        //new webpack.optimize.UglifyJsPlugin(),
        new webpack.ContextReplacementPlugin(/moment[\/\\]locale$/, /(en|es)\.js/),
        new CleanWebpackPlugin(['dist'], {
            verbose: true,
            dry: false
        }),
        new AssetsPlugin({
            path: path.join(__dirname, 'dist')
        }),
        new BeepPlugin()
    ],
    devtool: false,
    ts: {
        //transpileOnly: true,
        compilerOptions: {
            "noEmit": false
        }
    }
}
```

1) I'll try, but this should affect the **typescript only** case as well. Shouldn't it? 
2) I've  `devtool: false`, that produces the same result as doing nothing at all. I've tried enabling it and then takes much longer (like **1 min** or so)
3) `tsc -v` **1.8.7**
4) `node -v`  **v4.2.6**

**Comment by s-panferov** [comment]
1) TSC behavior is different. If it sees `files` section in you tsconfig file, it doesn't try to search for them on file system. The loader has slightly different behavior, it will search anyway.

**Comment by s-panferov** [comment]
@olmobrutall could you please also share your `tsconfig.json`?

**Comment by olmobrutall** [comment]
I've updated to TS 1.8.9 and now it looks a little bit faster **32s**, then i changed back to 1.8.7 and still **32s** so maybe is unrelated.

Then I've checked tsconfig.json: 

```JSON
{
  "compilerOptions": {
    "target": "es5",
    "sourceMap": true,
    "module": "commonjs",
    "jsx": "react",
    "noEmit": true
  },
  "exclude": [ "node_modules" ]
}
```

So `exclude` is already set.
I've tried to set `sourceMap` to false and then I get error: 

```
ERROR in [default]
Option 'sourceRoot' cannot be specified without specifying option 'sourceMap'.
```

**Comment by s-panferov** [comment]
@olmobrutall is your source code public? Could you please share or send it somehow?

**Comment by olmobrutall** [comment]
Is this proyect: https://github.com/signumsoftware/southwind

**Comment by olmobrutall** [comment]
Its big, actually in this branch and folder: https://github.com/signumsoftware/southwind/tree/React/Southwind.React and most of the code is in git-submodules:

https://github.com/signumsoftware/framework/tree/8c67e5b7fc2e7cc1b72a5744af4463739ec6ca84/Signum.React

https://github.com/signumsoftware/extensions/tree/f4bce50a28816fff7ab27ae42eeabce4f2d8bbdb/Signum.React.Extensions

**Comment by olmobrutall** [comment]
It's a Framework and a sample application that will be cloned (http://www.signumsoftware.com/en/DuplicateApplication), that's why I want to optimize the experience so much

**Comment by olmobrutall** [comment]
Also **33s** with  
```
{
  "compilerOptions": {
    "target": "es5",
    "sourceMap": false,
    "sourceRoot": "",
    "module": "commonjs",
    "jsx": "react",
    "noEmit": true
  },
  "exclude": [ "node_modules" ]
}
```

**Comment by s-panferov** [comment]
@olmobrutall thanks, I'll try to debug it tomorrow (I have GMT+3 timezone).

**Comment by olmobrutall** [comment]
Sure, I appreciate your help. If you need anything I'm in gitter / skype with the same name

**Comment by olmobrutall** [comment]
As recomended by @Den-dp (https://gitter.im/s-panferov/awesome-typescript-loader?at=57097c05afe21a92130d167c) 

* I've tried using `ForkCheckerPlugin`. In the description looks similar to `transpileOnly` feature of `ts-loader` but there where no speed changes. 

* I tried calling `webpack --profile --display-modules` and I'm not able to interpret the results. I can paste it but without colors are very hard to read. 

* The I've tried to run `webpack --profile --json > stats.json` and analyzed in `http://webpack.github.io/analyse` (again). The hints indicate that there could be something related to bootstrap, that even if it's a dependency was not included in the **dependencies only** profile, so I played a little bit with it: 

![image](https://cloud.githubusercontent.com/assets/720953/14409418/1a375cf8-ff13-11e5-9684-2ad9057cdaf4.png)

Bootstrap css adds **1s** and botstrap less **3s**. Relevant but nothing to do with the **30s** that the typescript loaders are taking.


I've uploaded the stats.json with and without my Typescript code : 
[statsDepsOnly.json.txt](https://github.com/s-panferov/awesome-typescript-loader/files/211714/statsDepsOnly.json.txt)
[statsWithTS.json.txt](https://github.com/s-panferov/awesome-typescript-loader/files/211716/statsWithTS.json.txt)

I can not identify what [profiler key](https://github.com/webpack/docs/wiki/cli#the-timing-keys) is the actual time that the module took. Once the TS code is included, the dependencies also take 5-digit millisencond measures in `factory`, `building` and `dependencies`.

**Comment by s-panferov** [comment]
@olmobrutall sorry for the delay, I've tried to debug several times but without any success for now.

But I think I've found the origin of the error: when a file requires with this construction

```
require(["./Frames/PageFrame"],  cb)
```

It takes a long time for webpack to compile it. I'll try to continue the investigation.

**Comment by s-panferov** [comment]
@olmobrutall yes, I think I finally understand what is going on, fix is on the way.

**Comment by s-panferov** [comment]
@olmobrutall please try version `0.18.0-rc.0`. It is published on npm with `next` tag.

**Comment by s-panferov** [comment]
I need to go afk right now, but later I will write detailed explanation about the issue.

**Comment by olmobrutall** [comment]
Hi again!

> But I think I've found the origin of the error: when a file requires with this construction
> ```require(["./Frames/PageFrame"],  cb)```
> It takes a long time for webpack to compile it. I'll try to continue the investigation.

I was suspecting something like this, that's why I tried using sync `require` instead but there were no important differences. 

I've run the benchmark again, now with a little bit more code. Here are the results.

![image](https://cloud.githubusercontent.com/assets/720953/14768057/bb218b60-0a36-11e6-8ae1-3034b157a817.png)


* there is a real improvement of around 16s using the new version for real TS compilations. 
* `ts-loader transpileOnly` is still the faster webpack solution 
* And TS by itself is way faster :S

**Comment by s-panferov** [comment]
@olmobrutall on your example codebase right now I have `10714ms build modules` and total time `14185ms`.

Could you publish your updated database?

**Comment by olmobrutall** [comment]
The latest version of southwind in React branch and Southwind.React folder is what I'm using.

just enable awesome-typescript-loader in webpack.config. 

Could you try any other combination to check of the differences are not due to he computer? Mine is a i7 16Gb with SSD so you could not be so much better.

**Comment by s-panferov** [comment]
@olmobrutall what version of NodeJS are you using?

**Comment by s-panferov** [comment]
And the second question, what is your OS?

**Comment by olmobrutall** [comment]
Node v4.4. Maybe is that?

 I had a problem with 5 but maybe is gone now

**Comment by olmobrutall** [comment]
Windows 10

**Comment by s-panferov** [comment]
@olmobrutall I'm using Mac Book 2,5 GHz Intel Core i7 with SSD, OSX operation system. I'll try to debug on Windows tomorrow, maybe I'll see the gap.

**Comment by olmobrutall** [comment]
I'll try with node 5 when I'm at home. Thanks a lot for tie help!!

**Comment by olmobrutall** [comment]
No differences with Node 5:

![image](https://cloud.githubusercontent.com/assets/720953/14769600/2dd22118-0a5c-11e6-8d44-88d2c5942c03.png)

**Comment by ConneXNL** [comment]
Is there a way to built in  a transpileOnly  in this library too ?

I can't seem to get fast build speeds without transpileOnly: true.

**Comment by omouse** [comment]
I just discovered HappyPack for Webpack which will use threads to speed things up, not sure if it's applicable here since it looks like the loader itself is slow: https://github.com/amireh/happypack

**Comment by faceyspacey** [comment]
how is this looking now? How does it compare to standard webpack compilation times?

**Comment by olmobrutall** [comment]
I manage to make it much faster in watch by using DLLPlugin. Very recommended, is worth the setup time.

More info here:

https://github.com/signumsoftware/framework/commit/970bb6a79baa5f5a02d76706a020181fffc3e532

**Comment by chrisdavies** [comment]
I'm seeing similar issues on my current build:

`time tsc` shows roughly 5 seconds.

With awesome-typescript-loader, it's 64 seconds.

Any tips on speeding things up?

Unfortunately, my code is closed-source.

tsconfig, if that's helpful:

```
{
  "compilerOptions": {
    "outDir": "./public/assets",
    "sourceMap": true,
    "noImplicitAny": true,
    "module": "commonjs",
    "target": "es5",
    "jsx": "react",
    "noUnusedLocals": true,
    "allowJs": true,
    "lib": [
      "es5",
      "es6",
      "dom"
    ]
  },
  "include": [
    "./js/**/*"
  ],
  "exclude": [
    "node_modules"
  ]
}
```

**Comment by chrisdavies** [comment]
OK. I had exclude in my `tsconfig`, but I didn't have `exclude: /node_modules/,` in my `webpack.config`.  Adding that brought the build down to 11 seconds, which is just fine with me!

**Comment by vitalybe** [comment]
@chrisdavies That helped me, thanks, this is what I did in my Webpack config:

          {
            test: /\.(t|j)sx?$/, exclude: /node_modules/, use: { loader: 'awesome-typescript-loader',  }
          },


My compilation time went from 75 seconds to 5 seconds. Is that the right way however? I feel like I am doing something wrong. Would the node_modules still be included in the final bundle?

I've originally followed the conversion guide here and it doesn't mention exclusion: https://github.com/Microsoft/TypeScript-React-Conversion-Guide

Hmm... Come to think of it for Babel we've also used to exclude node_modules, so that makes sense. If so, perhaps should make an issue there.
